### PR TITLE
Fix averaging phasor coordinates and related issues

### LINKED ISSUE #122: Improve `phasor_threshold` to handle multiple harmonics simultaneously.
Current implementation of `phasor_threshold` exploits numpy broadcasting. This is not optimal in the case where multiple harmonics are present, since `mean` and `real` and `imag` have different dimensions (`mean` doesn't have the harmonics dimension) and the product of thresholding all harmonics at once is a new dimension for `mean` for  the harmonic, which from the theoretical point of view, doesn't make sense.

Also, `NaN` values should be propagated to all harmonics and components, which is not being done currently by `phasor_threshold`

For now, processing harmonics separately is recommended to avoid this problems, until a solution is implemented.

### PR DESCRIPTION
**Pull request by cgohlke**
## Description

This PR proposes several new features and breaking changes to fix issues raised in #99 and #122:

- Add `normalize` parameter to `phasor_from_signal` function.

  This is a trivial change but might be useful for experimenting with filter algorithms.

- Add `phasor_normalize` function.

  Use to normalize phasor coordinates returned by `phasor_from_signal(..., normalize=False)` or provided by user.

  Do we need a `phasor_unnormalize` function, maybe for internal use? It's probably a good idea that no other public function should require unnormalized coordinates. Un/re-normalization can be transparently handled internally if necessary.

- Rename `phasor_filter` function to `phasor_filter_median`:

  - Remove `method` parameter.
  - Add `use_scipy` parameter to be used instead of `method='median_scipy'`.
  - Add `mean` parameter to detect presence of harmonic axis in `real` and `imag` and make the API like other functions.
  - Remove `'axes'` from `kwargs`. It's redundant to `skip_axis`.
  - Return `mean` input unchanged to make API match other functions.
  - Remove private `_median_filter_scipy` and `_median_filter` functions. Their implementation moved into `phasor_filter_median`.

- Support multiple harmonics in `phasor_threshold`:

  - Propagate NaNs in `mean` or any `real` and `imag` harmonics to `mean` and all harmonics in `real` and `imag`.
  - Add `detect_harmonics` option to skip detecting multiple harmonics and revert to pure ufunc behavior.

- Fix `phasor_center` function:

  - Add `mean` parameter to fix averaging phasor coordinates, detect presence of harmonic axis in `real` and `imag`, and make API similar with other functions.
  - Use `phasor_threshold` for `method='mean'` to propagate NaNs among arrays to ensure the averages are calculated over the same elements. Should this be done also for `method='median'`?
  - Return filtered `mean` array, required to correctly interpret averaged phasor coordinates and make the API like other functions.
  - Change `skip_axis` refer to axes in `mean`, not `real`. The harmonic axis, if present, is automatically excluded.

- Fix `phasor_calibrate` function:

  - Add `reference_mean` parameter, which is required by the `phasor_center` function and used to automatically detect presence of harmonics.
  - Change `skip_axis` refer to axes in `reference_mean`, not `reference_real`. The harmonic axis, if present, is automatically excluded.

- Remove formal docstrings from private `_mean` and `_median` functions.

  They are repetitive, distracting, and not really needed to understand the functions.

- Adjust tests and tutorials for the above changes and features:

  Multi-harmonic support in `phasor_calibrate` and `phasor_filter_median` make the tutorials cleaner.

Closes #122.

## Checklist

- [x] The pull request title and description are concise.
- [x] Related issues are linked in the description.
- [x] New dependencies are explained.
- [x] The source code and documentation can be distributed under the [MIT license](https://www.phasorpy.org/docs/stable/license/).
- [x] The source code adheres to [code standards](https://www.phasorpy.org/docs/stable/contributing/#code-standards).
- [x] New classes, functions, and features are thoroughly [tested](https://www.phasorpy.org/docs/stable/contributing/#tests).
- [x] New, user-facing classes, functions, and features are [documented](https://www.phasorpy.org/docs/stable/contributing/#documentation).
- [x] New features are covered in tutorials.
- [x] No files other than source code, documentation, and project settings are added to the repository.

### CODE REVIEW

**Comment by bruno-pannunzio** [comment]
Hi @cgohlke, sorry I wasn't able to respond before, it has been crazy this past week with the workshop.

I can review this in detail tomorrow or Friday since today is the last day of the course.

File: src/phasorpy/_phasorpy.pyx [hunk 1]
```diff
@@ -924,35 +933,72 @@ cdef (float_t, float_t) _phasor_at_harmonic(
...
+#     float_t dividend,
+#     float_t divisor,
+# ) noexcept nogil:
+#     """Return dividend divided by divisor."""
+#     if divisor != 0.0:
+#         # includes isnan(divisor)
+#         return dividend / divisor
+#     return NAN if dividend == 0.0 else dividend * INFINITY
+
+
```
Comment (bruno-pannunzio) [inline src/phasorpy/_phasorpy.pyx:1001]: Is it intended to keep this lines commented?

**Comment by bruno-pannunzio** [comment]
Hi @cgohlke! I think all of this are great improvements to the workflow. Particularly all the tutorials are much clear now that all functions can handle harmonics. Regarding this, I also see that in the basic workflow for phasor analysis, the `plot_phasor` is the only one that doesn't support multiple harmonics and each harmonic has to be plotter separately. Do you think at some point it would make sense to modify the implementation of `plot_phasor`, or better yet the plot methods in `PhasorPlot`, to detect and plot harmonics separately? Plotting multiple harmonics in the same plot doesn't make much sense by default and it would be nice if it would support this detection of harmonics and plotting all separately. But I don't know how much work needs to be put into that.

I tested all the modifications and the results for multiple-harmonic analysis are the same as before, so everything seems to work as expected. As I said before, now the workflow is much easier not having to keep track of the harmonic axis and all related.

 > * Do we need a `phasor_unnormalize` function, maybe for internal use? It's probably a good idea that no other public function should require unnormalized coordinates. Un/re-normalization can be transparently handled internally if necessary.

I was going to propose this exactly. If it's not much trouble, it would be very nice to have this also because we are going to need this for other functions. I think it's a good idea to keep it private, at least for now.

File: src/phasorpy/_phasorpy.pyx (hunk 1 above)
Comment (cgohlke) [reply src/phasorpy/_phasorpy.pyx:1001]: These lines can be removed. I needed to check that numpy divide ufuns match the Cython code we use elsewhere.

**Comment by cgohlke** [comment]
> Do you think at some point it would make sense to modify the implementation of plot_phasor, or better yet the plot methods in PhasorPlot, to detect and plot harmonics separately?

`PhasorPlot` is intrinsically single harmonic. `plot_phasor` could be changed to plot a grid of single-harmonic phasor plots. Currently it can't automatically detect the presence of harmonics. We would need to add a parameter and/or pass `mean`. Anyway, that would go into a separate PR.

> If it's not much trouble, it would be very nice to have this also because we are going to need this for other functions. I think it's a good idea to keep it private, at least for now.

I'll add a private function. We'll see how useful it is. I would expect the un/normalization to be done more efficiently "inline", like in the `_mean` function, or directly in a Cython/numpy ufunc.

>> Use phasor_threshold for `method='mean'` to propagate NaNs among arrays to ensure the averages are calculated over the same elements. Should this be done also for `method='median'`?

Any thoughts on this?

**Comment by bruno-pannunzio** [comment]
> * Use `phasor_threshold` for `method='mean'` to propagate NaNs among arrays to ensure the averages are calculated over the same elements. Should this be done also for `method='median'`?

I think using `phasor_threshold` for propagating NaNs is a clever way to do this. Also I think this is not computationally too heavy so it should add too much to the function. I liked how this is done. I think it can be useful also for the `median` method, but maybe I am missing something on why it should be only used with the `mean` method.

**Comment by cgohlke** [comment]
> I think it can be useful also for the median method, but maybe I am missing something on why it should be only used with the mean method.

How about adding a `threshold : bool | None = None` parameter such that the user can explicitly decide if necessary (or not, for example, if the input is already thresholded)? By default, it would be enabled for `method='mean'`.

**Comment by bruno-pannunzio** [comment]
> How about adding a `threshold : bool | None = None` parameter such that the user can explicitly decide

This can be a good idea to optimize for cases not needed. Although maybe there is a more intuitive name for the parameter? Because although it uses the `phasor_threshold` the idea is to propagate the NaNs, so maybe something more related to the use of the parameter? But `threshold` can work as long as the description fits.

> By default, it would be enabled for `method='mean'`

I think it's a good idea to keep it by default, but I am missing something on why it should not be default for all methods but only for mean.

**Comment by cgohlke** [comment]
> Although maybe there is a more intuitive name for the parameter? 

Agreed. How about:

```
nan_safe: bool = True,

"""
nan_safe : bool, optional
    Ensure `method` is applied to same elements of input arrays.
    By default, distribute NaNs among input arrays before applying
    `method`.
"""
```

Other suggestions?

> Why it should not be default for all methods but only for mean.

Because it is not strictly necessary for median? And in another issue you argued not using thresholds before median filters (let's get back to that in Monday's meeting). Anyway, I'm in favor of applying `nan_safe=True` as a default.

**Comment by bruno-pannunzio** [comment]
> nan_safe: bool = True,
> 
> """
> nan_safe : bool, optional
>     Ensure `method` is applied to same elements of input arrays.
>     By default, distribute NaNs among input arrays before applying
>     `method`.
> """

I like this very much! This is clear that is NaN related and the description matches.

> Because it is not strictly necessary for median? And in another issue you argued not using thresholds before median filters (let's get back to that in Monday's meeting).

Oh you are totally right! I think I am also inclined to using it by default. I don't know the reason behind applying the threshold before filtering, it's a good idea to discuss and try to find the reason.

**Comment by cgohlke** [comment]
> If it's not much trouble, it would be very nice to have this also because we are going to need this for other functions.

I'll leave this out of this PR. Normalization is not completely reversible. If one really needs "un-normalized" coordinates, calculate them from the signal.

**Review by bruno-pannunzio — APPROVED** [review]
Great @cgohlke! I think this is ready considering what we discussed. 

Testing this for the wavelet filtered I encountered the same, unnormalizing after normalization is not so straight-forward. Also, I am not sure all functions work well with unnormalized data. For example, I tried the basic pipeline and I think `phasor_calibrate` is not working correctly, but I need further testing. Maybe we can consider this at least for the most commonly used functions, but we can talk about this in the meeting.

**Pull request merged**